File: ios/SecureClipboard.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(SecureClipboard, NSObject)

RCT_EXTERN_METHOD(setString:(NSString *)text
                  expirationMs:(NSNumber *)expirationMs
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getString:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(clearString:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/SecureClipboard.swift
import Foundation
import React
import UniformTypeIdentifiers

@objc(SecureClipboard)
class SecureClipboard: NSObject, RCTBridgeModule {
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  @objc
  static func moduleName() -> String! {
    return "SecureClipboard"
  }
  
  @objc
  func setString(_ text: String, expirationMs: NSNumber, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      let pasteboard = UIPasteboard.general
      
      // Always treat data as sensitive for secure clipboard service
      let expirationSeconds = Double(expirationMs.intValue) / 1000.0
      
      // Set expiration time using native iOS pasteboard expiration (iOS 15.1+)
      if expirationSeconds > 0, #available(iOS 15.1, *) {
        pasteboard.setItems([[UTType.utf8PlainText.identifier: text]], options: [
          UIPasteboard.OptionsKey.expirationDate: Date().addingTimeInterval(expirationSeconds)
        ])
      } else {
        // No expiration or older iOS version, just set the text
        pasteboard.string = text
      }
      
      resolver(nil)
    }
  }
  
  @objc
  func getString(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      let pasteboard = UIPasteboard.general
      let text = pasteboard.string ?? ""
      resolver(text)
    }
  }
  
  @objc
  func clearString(_ resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      let pasteboard = UIPasteboard.general
      pasteboard.string = ""
      resolver(nil)
    }
  }
}
